tests/testsuite/rustflags.rs

```diff
@@ -993,6 +1054,188 @@ fn target_rustflags_normal_source() {
         .run();
 }
 
+#[cargo_test]
+fn target_rustflags_also_for_build_scripts() {
+    let p = project()
+        .file("src/lib.rs", "")
+        .file(
+            "build.rs",
+            r#"
+                fn main() { }
+                #[cfg(not(foo))]
+                fn main() { }
+            "#,
+        )
+        .file(
+            ".cargo/config",
+            &format!(
+                "
+            [target.{}]
+            rustflags = [\"--cfg=foo\"]
+            ",
+                rustc_host()
+            ),
+        )
+        .build();
+
+    p.cargo("build").run();
+}
+
+#[cargo_test]
+fn target_rustflags_not_for_build_scripts_with_target() {
+    let host = rustc_host();
+    let p = project()
+        .file("src/lib.rs", "")
+        .file(
+            "build.rs",
+            r#"
+                fn main() { }
+                #[cfg(foo)]
+                fn main() { }
+            "#,
+        )
+        .file(
+            ".cargo/config",
+            &format!(
+                "
+            [target.{}]
+            rustflags = [\"--cfg=foo\"]
+            ",
+                host
+            ),
+        )
+        .build();
+
+    p.cargo("build --target").arg(host).run();
+
+    // Enabling -Ztarget-applies-to-host should not make a difference without the config setting
+    p.cargo("build --target")
+        .arg(host)
+        .masquerade_as_nightly_cargo()
+        .arg("-Ztarget-applies-to-host")
+        .run();
+
+    // Even with the setting, the rustflags from `target.` should not apply, to match the legacy
+    // Cargo behavior.
+    p.change_file(
+        ".cargo/config",
+        &format!(
+            "
+        target-applies-to-host = true
+
+        [target.{}]
+        rustflags = [\"--cfg=foo\"]
+        ",
+            host
+        ),
+    );
+    p.cargo("build --target")
+        .arg(host)
+        .masquerade_as_nightly_cargo()
+        .arg("-Ztarget-applies-to-host")
+        .run();
+}
+
+#[cargo_test]
+fn build_rustflags_for_build_scripts() {
+    let host = rustc_host();
+    let p = project()
+        .file("src/lib.rs", "")
+        .file(
+            "build.rs",
+            r#"
+                fn main() { }
+                #[cfg(foo)]
+                fn main() { }
+            "#,
+        )
+        .file(
+            ".cargo/config",
+            "
+            [build]
+            rustflags = [\"--cfg=foo\"]
+            ",
+        )
+        .build();
+
+    // With "legacy" behavior, build.rustflags should apply to build scripts without --target
+    p.cargo("build")
+        .with_status(101)
+        .with_stderr_does_not_contain("[..]build_script_build[..]")
```

Comment:
Could this be a more targeted test and assert that the compilation of the build script gets `--cfg=foo` as expected?

Reply:
Done!